d2compiler/compile.go

```diff
@@ -27,18 +29,46 @@ type CompileOptions struct {
 	FS fs.FS
 }
 
+// Changes for Language Server 'mode'
+type LspOutputData struct {
+	Ast *d2ast.Map
+	Err error
+}
+
+var lod LspOutputData
+
+func LspOutput(m bool) {
+	if !m {
+		return
+	}
+
+	jsonOutput, _ := json.Marshal(lod)
+	fmt.Print(string(jsonOutput))
+	os.Exit(42)
+}
+
 func Compile(p string, r io.Reader, opts *CompileOptions) (*d2graph.Graph, *d2target.Config, error) {
 	if opts == nil {
 		opts = &CompileOptions{}
 	}
 
+	lspMode := os.Getenv("D2_LSP_MODE") == "1"
+	defer LspOutput(lspMode)
+
 	ast, err := d2parser.Parse(p, r, &d2parser.ParseOptions{
 		UTF16Pos: opts.UTF16Pos,
 	})
+
+	lod.Ast = ast
 	if err != nil {
+		lod.Err = err
 		return nil, nil, err
 	}
 
+	if lspMode {
+		return nil, nil, nil
+	}
```

Comment:
```suggestion
func Compile(p string, r io.Reader, opts *CompileOptions) (*d2graph.Graph, *d2target.Config, error) {
	if opts == nil {
		opts = &CompileOptions{}
	}

	ast, err := d2parser.Parse(p, r, &d2parser.ParseOptions{
		UTF16Pos: opts.UTF16Pos,
	})

	if os.Getenv("D2_LSP_MODE") == "1" {
		jsonOutput, _ := json.Marshal(LspOutputData{Ast: ast, Err: err})
		fmt.Print(string(jsonOutput))
		os.Exit(42)
		return nil, nil, err
	}

	if err != nil {
		return nil, nil, err
	}
```

I think it is cleaner to do this without the global + defer

Reply:
Yes, better.  I didn't pull off my goal of least amount of code churn possible.